temporal-sdk/src/test/java/io/temporal/workflow/updateTest/UpdateWithStartTest.java

```diff
@@ -0,0 +1,484 @@
+
+    try {
+      WorkflowClient.updateWithStart(workflow::update, 0, "Hello Update", updateOp);
+      fail("unreachable");
+    } catch (IllegalArgumentException e) {
+      assertEquals(e.getMessage(), "Method 'update' is not a WorkflowMethod");
+    }
+
+    ensureNoWorkflowStarted(workflowClient, options.getWorkflowId());
+  }
+
+  @Test
+  public void failWhenMixingStubs() {
+    WorkflowClient workflowClient = testWorkflowRule.getWorkflowClient();
+
+    WorkflowOptions options = createOptions();
+    TestWorkflows.TestUpdatedWorkflow stub1 =
+        workflowClient.newWorkflowStub(TestWorkflows.TestUpdatedWorkflow.class, options);
+    UpdateWithStartWorkflowOperation<Void> updateOp =
+        UpdateWithStartWorkflowOperation.newBuilder(stub1::update, "Hello Update")
+            .setWaitForStage(WorkflowUpdateStage.ACCEPTED)
+            .build();
+
+    TestWorkflows.WorkflowWithUpdate stub2 =
+        workflowClient.newWorkflowStub(TestWorkflows.WorkflowWithUpdate.class, options);
+    try {
+      WorkflowClient.updateWithStart(stub2::execute, updateOp);
+      fail("unreachable");
+    } catch (IllegalArgumentException e) {
+      assertEquals(
+          e.getMessage(), "UpdateWithStartWorkflowOperation invoked on different workflow stubs");
+    }
+
+    ensureNoWorkflowStarted(workflowClient, options.getWorkflowId());
+  }
+
+  private static void ensureNoWorkflowStarted(WorkflowClient workflowClient, String workflowId) {
+    try {
+      workflowClient.fetchHistory(workflowId);
+      fail("unreachable");
+    } catch (StatusRuntimeException e) {
+      assertEquals(e.getStatus().getCode(), Status.NOT_FOUND.getCode());
+    }
+  }
+
+  private WorkflowOptions createOptions() {
+    return SDKTestOptions.newWorkflowOptionsWithTimeouts(testWorkflowRule.getTaskQueue())
+        .toBuilder()
+        .setWorkflowId(UUID.randomUUID().toString())
+        .build();
+  }
+
+  public static class WorkflowWithUpdateImpl implements TestWorkflows.WorkflowWithUpdate {
+    String state = "initial";
+    CompletablePromise<Void> promise = Workflow.newPromise();
+
+    @Override
+    public String execute() {
+      promise.get();
+      return state;
+    }
+
+    @Override
+    public String getState() {
+      return state;
+    }
+
+    @Override
+    public String update(Integer index, String value) {
+      state = value;
+      return value;
+    }
+
+    @Override
+    public void updateValidator(Integer index, String value) {
+      if (index < 0) {
+        throw new RuntimeException("Rejecting update");
+      }
+    }
+
+    @Override
+    public void complete() {
+      state += " complete";
+      promise.complete(null);
+    }
+
+    @Override
+    public void completeValidator() {}
+  }
+
+  public static class TestUpdatedWorkflowImpl implements TestWorkflows.TestUpdatedWorkflow {
+
+    @Override
+    public String execute() {
+      return "done";
+    }
+
+    @Override
+    public void update(String arg) {}
```

Comment:
Can we test calling this as part of an update with start successfully as well? I see we use it in a test, but I don't think we actually test it can work. The reason I want to test it is because it returns a `void` type and in and sometimes that can cause problems.

Reply:
👍 good idea! done: https://github.com/temporalio/sdk-java/pull/2199/commits/7ed0ddd439bbb074d940f838eae711d178c0221b